generation/langchain_single_pass/generator_langchain.py

```diff
@@ -0,0 +1,799 @@
+    mode = "a" if append else "w"
+    with open(iteration_log_path, mode, encoding="utf-8") as log_file:
+        log_file.write(formatted_output)
+    
+    # Return the data dictionary for potential further processing
+    return iteration_data
+
+
+
+
+
+
+
+def generate_kernel_with_direct_docs_and_error_loop(
+    system_prompt_path, 
+    user_prompt_path, 
+    output_address,
+    kernel_module_path,
+    test_script_path,
+    test_script_output,
+    reasoning_log_path,
+    error_doc_path,
+    docs_dir,
+    max_iterations=15
+):
+    """
+    Generate a NKI kernel using direct function documentation access and iteratively 
+    improve it based on error feedback with detailed error documentation.
+    """
+    print("Initializing components...")
+    
+    # Initialize the error parser
+    print(f"Initializing NKI error parser from {error_doc_path}")
+    error_parser = NKIErrorParser(error_doc_path)
+    print(f"Loaded {len(error_parser.list_all_errors())} error codes from documentation")
+    
+    # Set up detailed trace log file
+    trace_log_path = output_address + ".detailed_trace.txt"
+    log_to_file(trace_log_path, "=== DETAILED TRACE LOG ===", append=False)
+    log_to_file(trace_log_path, f"Starting new kernel generation process at {datetime.datetime.now()}")
+    
+    # Set up consolidated iteration log file
+    consolidated_log_path = output_address + ".consolidated_iterations.txt"
+    # Initialize with header only on first write (will be overwritten)
+    with open(consolidated_log_path, "w", encoding="utf-8") as f:
+        f.write(f"=== CONSOLIDATED ITERATION LOG ===\n")
+        f.write(f"Started at: {datetime.datetime.now()}\n")
+        f.write(f"Output path: {output_address}\n")
+        f.write(f"Kernel module path: {kernel_module_path}\n\n")
+
+    # Load the initial prompts
+    system_prompt = read_file(system_prompt_path)
+    user_prompt = read_file(user_prompt_path)
+    
+    log_to_file(trace_log_path, f"System Prompt:\n{system_prompt}\n")
+    log_to_file(trace_log_path, f"User Prompt:\n{user_prompt}\n")
+    
+    print(f"Starting documentation-based generation for: {user_prompt[:50]}...")
+    
+    # Initialize LLMs
+    query_llm = ChatOpenAI(
+        model="gpt-4o-mini",
+        temperature=0.3
+    )
+    
+    # kernel_llm = ChatOpenAI(
+    #     model="gpt-4o-mini", 
+    #     temperature=0.85
+    # )
+    kernel_llm = ChatBedrock(
+        model_id="anthropic.claude-3-5-sonnet-20241022-v2:0",
+        model_kwargs={"temperature": 0.85},  # Move temperature into model_kwargs
+        region_name="us-west-2"
+    )
+
+
+    # Get list of available functions
+    available_functions = get_available_functions(docs_dir)
+    print(f"Found {len(available_functions)} available NKI functions in documentation")
+    log_to_file(trace_log_path, f"AVAILABLE FUNCTIONS:\n{', '.join(available_functions)}\n")
+    
+    # Initial kernel generation with direct documentation
+    try:
```

Comment:
If this works on your end, very cool! You are probably already thinking this, but in your next pass on this can you try to clean up this function so it's more modular? Having really long processes like this is really challenging, not maintainable, difficult to debug, etc.